forge/last-checker.rkt

```diff
@@ -9,91 +10,179 @@
   (prefix-in @ (only-in racket -> >=)))
 
 (provide checkFormula checkExpression primify)
+; This function only exists for code-reuse - it's so that we don't
+; need to use begin and hash-ref in every single branch of the match
+; Given a struct to handle, a hashtable to search for a handler in,
+; and an output to return, executes the check in the handler
+; then returns the output
+(define (check-and-output ast-node to-handle checker-hash output)
+  (begin (when (hash-has-key? checker-hash to-handle) ((hash-ref checker-hash to-handle) ast-node))
+         output))
+
 
 ; Recursive descent for last-minute consistency checking
-; Catch potential issues not detected by AST construction + generate warnings 
+; Catch potential issues not detected by AST construction + generate warnings
+(define/contract (checkFormula run-or-state formula quantvars checker-hash)
+  (@-> (or/c Run? State? Run-spec?)
+       (or/c node/formula? node/expr?)
+       list?
+       hash?
+       (or/c boolean? void? (listof (listof symbol?))))
 
-(define (checkFormula run-or-state formula quantvars)
   (when (@>= (get-verbosity) VERBOSITY_DEBUG)
     (printf "last-checker: checkFormula: ~a~n" formula))
   
   (match formula    
     [(node/formula/constant info type)
-     #f]
+     (check-and-output formula node/formula/constant checker-hash #f)]
     
     [(node/formula/op info args)
-     (checkFormulaOp run-or-state formula quantvars args)]
+     (check-and-output formula
+                       node/formula/op
+                       checker-hash
+                       (checkFormulaOp run-or-state formula quantvars args checker-hash))]
```

Comment:
For all these calls to `(check-and-output A B C D)`, are we sure the order of side effects is ok? Is it safe to evaluate `D` before running the `checker-hash` checks?

I'd find these all easier to read as `(check-and-output A B C) D`

Reply:
agree